Feature/vectorization

rstub opened the pull request:
addresses #13 

For the other functions like `swe_azalt`, `swe_pheno_ut` etc. I am unsure if it makes sense to vectorize them. What do you think, @vreijs?

rstub committed:
Vectorize swe_calc

At the C++ level we expect equal length vectors for bodies and dates.
The R wrapper for the C++ function is not exported. Instead a manual
R wrapper is provided, which handles the cases where either one body
and many dates or one date and many bodies are supplied.

rstub committed:
Vectorize swe_deltat()

rstub committed:
Vectorize swe_fixstar2()

rstub committed:
Vectorize swe_fixstar2_mag()

rstub commented:
This is weird. The CI results from Appveyor and Travis show that on Linux I get the same deltat values as I got locally. But on [Mac OS](https://travis-ci.org/rstub/swephR/jobs/424517017#L412) and [Windows](https://ci.appveyor.com/project/rstub/swephr/build/1.0.69#L309), different deltat values are produced. Note that the values produced on Mac OS and Windows are identical.

@vreijs Any idea what might be the issue here?

rstub committed:
Add missing docu change for swe_fixstar()

rstub committed:
Vectorize swe_deltat_ex()

rstub commented:
It seems as if swephRdata is installed on Linux but not on Mac OS. I have not checked Windows yet. The deltat values might correspond to those with the build in ephemeris.

vreijs commented:
Where is this data installed? I also get all these error when running the
test. E.g.:

test-misc.R:109: skip: Moon near present day with SE
swephRdata cannot be loaded


I also see in the test-misc.R no load of this data.

I also issued a bug report to the swisseph e-group. It looks that the
swe_set_topo() info is not kept after a swe_close() (contrarily to age 6 f
the manual: swephprg.docs).

vreijs commented:
On 4 September 2018 at 21:59, Ralf Stubner <notifications@github.com> wrote:

> For the other functions like swe_azalt, swe_pheno_ut etc. I am unsure if
> it makes sense to vectorize them. What do you think, @vreijs
> <https://github.com/vreijs>?
>

In ARCHAEOCOSMO I just vectorised most functions. As I am not fully sed to
that concept I might not see it as important, but I think in the future
vectorisiong is just so handy. So I think we should do  most functions. In
ARCHAEOCOSMO I have single value input function as S_<name> and vectorised
input function as <name>.

vreijs commented:
On 4 September 2018 at 22:10, Ralf Stubner <notifications@github.com> wrote:

> This is weird. The CI results from Appveyor and Travis show that on Linux
> I get the same deltat values as I got locally. But on Mac OS
> <https://travis-ci.org/rstub/swephR/jobs/424517017#L412> and Windows
> <https://ci.appveyor.com/project/rstub/swephr/build/1.0.69#L309>,
> different deltat values are produced. Note that the values produced on Mac
> OS and Windows are identical.
>

I see also different DeltaT values then your test_misc.R values:


test-misc.R:25: failure: deltat can be retrieved
swe_deltat(1234.567) not equal to 1.597676.
1/1 mismatches
[1] 1.59 - 1.6 == -0.0103
test-misc.R:29: failure: deltat can be retrieved for vector
swe_deltat(c(1234.567, 1234567)) not equal to c(1.5976757, 0.3685434).
2/2 mismatches (average diff: 0.0064)
[1] 1.587 - 1.598 == -0.0103
[2] 0.366 - 0.369 == -0.0025


Do you see the same as I under Windows? I think this is also something
we also need to mention on the swisseph e-group.

All the best,


Victor

vreijs commented:
I think added swe_set_ephe_path("") is important, see the remark on.
. page 5 (first function to be called)
. page 41 (deltat)
. page >>> (calc[_ut])
. on page 75 it seems that some errors were removed when not calling
swe_set_ ephe_path, perhaps we can provide more input on 'errors' when not
calling it.

Could this also cause the difference in DeltaT value sin UNIX and windows?
So I have the following statements at the start of test-misc.R:
swe_set_ephe_path('')
swe_set_topo(0,50,10)
#the set_topo function seems not to survive close() so it needs to be
repeated, hopefully in next SE version this is rectified

By the way, can we make swe_set_ephe_path(NULL) to work? I think this needs
a if function in the call function.

<there is a methodology in SE to find the directory, I can't find that
anymore, that ingo might be important for where the swephR's sefstars.txt
is located>

rstub committed:
Take the used ephemeris into account when testing swe_deltat

rstub committed:
Inhibit overloading of ephemeris from pacakges

rstub commented:
I have found the source of these errors, although I cannot fully explain how we arrived in this situation. For some reason, the Travis Linux machines have the `swephRdata` package installed, while the Travis Mac OS machines and Appveyor Windows machines do not have that package installed. In addition, I have the package installed, while you don't have it.

This package, which can be found at https://github.com/rstub/swephRdata, was introduced in #16. The main idea is to have one R package that contains nothing but the SE files. If this package is installed, `swephR` automatically has the SE files available. Otherwise only the `sefstars.txt` and the internal ephemeris can be used. 

How does that work? In https://github.com/rstub/swephR/blob/master/R/zzz.R the `.onLoad` function is defined, which is automatically executed when the package is loaded:

```
.onLoad <- function(libname, pkgname) {
    if(requireNamespace("swephRdata", quietly = TRUE)) {
        swe_set_ephe_path(system.file("ephemeris", package = "swephRdata"))
    } else {
        swe_set_ephe_path(file.path(libname, pkgname, "ephemeris"))
    }
}
```

If the `swephRdata` package is available, the ephemeris path is set to the `ephemeris` sub-directory of the installation location of the  `swephRdata` package. Otherwise the `ephemeris` sub-directory of the installation location of the  `swephR` package is used. The former contains all the SE files, the latter only `sefstars.txt`. This way the user can decide to either save space and use the build in approximation or to install the add-on package gaining speed and precision. If one wants to go even further and add the full JPL, there are two possibilities:

1. Manually call `swe_set_ephe_path("...")` with the correct path or an empty path if the files are in a standard location.
2. Set the environment variable `SE_EPHE_PATH` to the desired location. If that variable is defined, the argument to `swe_set_ephe_path("...")` is ignored, c.f. page 5 in the programmers manual.

The second approach might be the best way for you. Even though I will have to unset that variable for the tests, since at least `swe_deltat` depends on the automatically chosen ephemeris in an uncontrollable way. Right now I have introduced a simple if

```
    if (requireNamespace("swephRdata"))
        expect_equal(swe_deltat(1234.567), 1.5976757, tolerance = .0000001)
    else
        expect_equal(swe_deltat(1234.567), 1.5873865, tolerance = .0000001)
```

If `swephRdata` is available, I expect the SE files to be used. If it is not available, I expect the internal ephemeris to be used. If it might happen, that `SE_EPHE_PATH` points to a directory with content I cannot control, then I cannot formulate any tests that I expect to hold on other systems (eventually CRAN).

BTW, the above if statement is only a workaround to deal with `swe_deltat`'s behavior that does not allow for actively controlling which ephemeris is used. For `swe_deltat_ex` it is much nicer as one can actively control which ephemeris should be used. Of course, in the formalized tests I cannot use the SE without the `swephRdata` package being installed. That's why these tests are skipped on your machine. That is totall fine and does not constitute an error.

More on the other points tomorrow. Now it is time to get some sleep ...

vreijs commented:
Hello Ralf,
Good the error is found. I think there are two statements missing at the
start of test-misc.R file:

swe_set_ephe_path('')
swe_set_topo(0,50,10)
# the set_topo should be needed here (but due to an error in the SE code,
it is now needed to be called before fixstar2 and calc(_ut) function)...

The only skips I now see in Windows10 are the missing swephRdata (I don't
want to download this package yet;-)


On 5 September 2018 at 22:28, Ralf Stubner <notifications@github.com> wrote:

> I have found the source of these errors, although I cannot fully explain
> how we arrived in this situation. For some reason, the Travis Linux
> machines have the swephRdata package installed, while the Travis Mac OS
> machines and Appveyor Windows machines do not have that package installed.
> In addition, I have the package installed, while you don't have it.
>
> This package, which can be found at https://github.com/rstub/swephRdata,
> was introduced in #16 <https://github.com/rstub/swephR/pull/16>. The main
> idea is to have one R package that contains nothing but the SE files. If
> this package is installed, swephR automatically has the SE files
> available. Otherwise only the sefstars.txt and the internal ephemeris can
> be used.
>
> How does that work? In https://github.com/rstub/swephR/blob/master/R/zzz.R
> the .onLoad function is defined, which is automatically executed when the
> package is loaded:
>
> .onLoad <- function(libname, pkgname) {
>     if(requireNamespace("swephRdata", quietly = TRUE)) {
>         swe_set_ephe_path(system.file("ephemeris", package = "swephRdata"))
>     } else {
>         swe_set_ephe_path(file.path(libname, pkgname, "ephemeris"))
>     }
> }
>
> If the swephRdata package is available, the ephemeris path is set to the
> ephemeris sub-directory of the installation location of the swephRdata
> package. Otherwise the ephemeris sub-directory of the installation
> location of the swephR package is used. The former contains all the SE
> files, the latter only sefstars.txt. This way the user can decide to
> either save space and use the build in approximation or to install the
> add-on package gaining speed and precision. If one wants to go even further
> and add the full JPL, there are two possibilities:
>
>    1. Manually call swe_set_ephe_path("...") with the correct path or an
>    empty path if the files are in a standard location.
>    2. Set the environment variable SE_EPHE_PATH to the desired location.
>    If that variable is defined, the argument to swe_set_ephe_path("...")
>    is ignored, c.f. page 5 in the programmers manual.
>
> The second approach might be the best way for you. Even though I will have
> to unset that variable for the tests, since at least swe_deltat depends
> on the automatically chosen ephemeris in an uncontrollable way. Right now I
> have introduced a simple if
>
>     if (requireNamespace("swephRdata"))
>         expect_equal(swe_deltat(1234.567), 1.5976757, tolerance = .0000001)
>     else
>         expect_equal(swe_deltat(1234.567), 1.5873865, tolerance = .0000001)
>
> If swephRdata is available, I expect the SE files to be used. If it is
> not available, I expect the internal ephemeris to be used. If it might
> happen, that SE_EPHE_PATH points to a directory with content I cannot
> control, then I cannot formulate any tests that I expect to hold on other
> systems (eventually CRAN).
>
> BTW, the above if statement is only a workaround to deal with swe_deltat's
> behavior that does not allow for actively controlling which ephemeris is
> used. For swe_deltat_ex it is much nicer as one can actively control
> which ephemeris should be used. Of course, in the formalized tests I cannot
> use the SE without the swephRdata package being installed. That's why
> these tests are skipped on your machine. That is totall fine and does not
> constitute an error.
>
> More on the other points tomorrow. Now it is time to get some sleep ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#issuecomment-418887443>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FsupOSh1jrRYCnkAH6YTC0j4xZTXks5uYEHsgaJpZM4WZw00>
> .
>

vreijs committed:
add thwo general command for test arund fixstar and calc

vreijs committed:
added heliecal_ut, but testing testheliecal.R causes instability of RStudio/R (it does a restart).

rstub commented on tests/testthat/test-misc.R:
```diff
@@ -1,5 +1,8 @@
 context("test-misc.R")
 
+swe_set_ephe_path('')
```
I am sorry, but this makes sense only on your machine where the SE files are installed in a directory that SE expects. That is not the case on my machine and more importantly will never be the case on the machines used for automatic testing (CRAN, Travis, Appveyor). I will revert this commit.

rstub commented on tests/testthat/test-misc.R:
```diff
@@ -1,5 +1,8 @@
 context("test-misc.R")
 
+swe_set_ephe_path('')
+swe_set_topo(0,50,10)
```
From my POV this is unnecessary. In all the tests that request topocentric positions, `swe_set_topo` is already used.

rstub commented on NAMESPACE:
```diff
@@ -1,26 +1,4 @@
-export(swe_set_tid_acc)
-export(swe_set_topo)
-export(swe_sol_eclipse_when_loc)
-export(swe_version)
```
Oops, removing all the exports is not a good idea. Please inspect the changes before you commit them.

rstub commented on R/testheliacal.R:
```diff
@@ -0,0 +1,18 @@
+#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.25),c(25,1),"venus",1,260)
+
+#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.25),c(25,1),"moon",3,260)
+
```
This file makes no sense in the `R` directory. I will put it into `inst/examples` for now. But it should probably be converted into a proper test.

rstub committed:
Revert "add thwo general command for test arund fixstar and calc"

This reverts commit b0f2efd5c1dd4d82888d6e6a43896d205036173e.

rstub committed:
Move testhelical.R to example directory

rstub committed:
Recreate NAMESPACE

rstub committed:
Use only the first three values of dret

rstub committed:
Bump version number

rstub committed:
Unset SE_EPHE_PATH before loading swephR

rstub commented:
I have not managed to get the tests for `swe_heliacal_ut` working reliably, but at least the other tests are working again.

vreijs commented:
> +swe_set_ephe_path('')
> +swe_set_topo(0,50,10)
>
> From my POV this is unnecessary. In all the tests that request topocentric
> positions, swe_set_topo is already used.
>
Officially, according to the manual (as I stated) one should not repeat the
swe_set_topo (as it should survive the swe_close() function. But it seems
indeed that close() removes the topo values, so indeed one needs to repeat
(erroneously) that statement. I just prepared the code when SE has
corrected this error (or changed its manual). As said I am trying to
discuss this at the swiiseph e-group.
I agree with the present error in SE, one needs ot repeat the topo
statement.

vreijs commented:
>
> +swe_set_ephe_path('')

I am sorry, but this makes sense only on your machine where the SE files
> are installed in a directory that SE expects. That is not the case on my
> machine and more importantly will never be the case on the machines used
> for automatic testing (CRAN, Travis, Appveyor). I will revert this commit.
>

In that case the R tests are not passed on my system (or any other system
that uses the default SE guidelines)! Remember the default directory search
method is default in SE, I am using that methodology (not another new
methodology). I think we should try to keep as close as possible to SE's
methodology.
By the way this statement is important as it (no matter where issued, as
long as it is issued at least one before a call to fixstar2/calc) as it set
certain values.

vreijs commented:
What should I have done? I am still too much a beginner, as I would not
know how to manipulate this;-) I did not explicitly do these commands... I
'just' run 'Clean and rebuilt'

I was already a little proud that I could commit and push to your branch...
That happened it little by accident (after listening to a few git/github
for beginners videos. Tell me what I might have done wrong, so I can learn.

On 7 September 2018 at 23:00, Ralf Stubner <notifications@github.com> wrote:

> *@rstub* commented on this pull request.
> ------------------------------
>
> In NAMESPACE
> <https://github.com/rstub/swephR/pull/17#discussion_r216097669>:
>
> > -export(swe_fixstar2)
> -export(swe_fixstar2_mag)
> -export(swe_get_tid_acc)
> -export(swe_heliacal_ut)
> -export(swe_lun_eclipse_how)
> -export(swe_lun_eclipse_when)
> -export(swe_lun_eclipse_when_loc)
> -export(swe_pheno_ut)
> -export(swe_rise_trans)
> -export(swe_rise_trans_true_hor)
> -export(swe_set_delta_t_userdef)
> -export(swe_set_ephe_path)
> -export(swe_set_tid_acc)
> -export(swe_set_topo)
> -export(swe_sol_eclipse_when_loc)
> -export(swe_version)
>
> Oops, removing all the exports is not a good idea. Please inspect the
> changes before you commit them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#pullrequestreview-153516003>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FiZGjBpp0fCRApWth0uhw7jgKKzLks5uYuxygaJpZM4WZw00>
> .
>

vreijs commented:
I made a proper test: I added it in test-misc.R (there the results are ok
and repeatable on my system). But when issuing the commands on their own it
is making errors on my system: that is why I did a separate test (which
should be removed when the results don't restart RStudio/R). I just wanted
to give you a sample to show it goes wrong (the test-miscR does not show
that;-). Will do such tests in the future in the inst/example
directory.Thanks for the advice.

On 7 September 2018 at 23:01, Ralf Stubner <notifications@github.com> wrote:

> *@rstub* commented on this pull request.
> ------------------------------
>
> In R/testheliacal.R
> <https://github.com/rstub/swephR/pull/17#discussion_r216097887>:
>
> > +options(digits=15)
> +
> +for (i in 1:5) {
> +#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.15),c(25,1),"sirius",1,260)
> +#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.15),c(25,1),"sirius",1,256)
> +#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.15),c(25,1),"mars",1,260)
> +a <- swe_heliacal_ut(1234567+i,c(0,50,10),c(1013.25,15,50,0.15),c(25,1),"mercury",1,260)
> +print (i)
> +print (a$dret[1:3])
> +}
> +
> +
> +#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.25),c(25,1),"venus",1,260)
> +
> +#swe_heliacal_ut(1234567,c(0,50,10),c(1013.25,15,50,0.25),c(25,1),"moon",3,260)
> +
>
> This file makes no sense in the R directory. I will put it into
> inst/examples for now. But it should probably be converted into a proper
> test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#pullrequestreview-153516309>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6Fs1whUYNbGfzQrEZz2sa9gv75kO8ks5uYuzMgaJpZM4WZw00>
> .
>

rstub committed:
Workaround for swe_heliacal_ut()

vreijs commented:
By the way the message i get when I run swe_heliacal_ut standalone (and not
in ARCHAECOSMO) I get: 'R Session Aborted, R encontered a fatal error. The
session was terminated.  'Start New Session?'"

At it is difficult to understand when it happens. If I run my test it goes
well, if I run it a second time, it runs well but after some seconds of
inactivity on console (in RStudio) I get this abort message. Something is
timing-out in the background? I see the CPU load stays high (after the
results have been printed) and then the abort message comes. This is
regarding of the number of dret cells I print.
In ARCHAEOCOSMO I only use one dret cell (I thought that using more and
also these weird (uninitialised as SE author admit) value ones might be the
reason, but not the case). So don't fully understand it.

vreijs commented:
I got all the errors back:-(
We really need to get this better. Again I would like to keep SE's methods
in place (as much as possible). As the ephemeris and star file need to be
in the same location, I am really wondering if we need to keep these files
outside the R install or use the recommended SE method. Remember that the
sefstars.txt can be changed by a user (as I am doing several times for my
research, but I know others also do). We need to keep these use-case in
tact as much as possible IMHO.
That is why I by the way use the empty swe_set_path method (should be NULL,
but that does not work yet). In that case SE has its own method of finding
the ephemeris/star files (so no specific strings needed in R).

rstub commented:
> I got all the errors back:-( 

Can you post these errors here or in an issue? Please format them with the "insert code" button ('<>') to make it readable.

> As the ephemeris and star file need to be in the same location, [...]

No. You can have multiple directories that are searched for these files. See `swe_set_ephe_path` in the programmers manual.

> I am really wondering if we need to keep these files outside the R install or use the recommended SE method. Remember that the sefstars.txt can be changed by a user (as I am doing several times for my research, but I know others also do). We need to keep these use-case in tact as much as possible IMHO.

Agreed, but we have to find a way that allows for automatic testing. If we ever want to commit this package to CRAN, then it is mandatory for it to build, install and run tests on a machine over which we have no control. That's why I am stressing these automated tests done by Travis & Appveyor so much.

rstub commented:
> What should I have done? I am still too much a beginner, as I would not know how to manipulate this;-) I did not explicitly do these commands... I 'just' run 'Clean and rebuilt'

Always look at the diff/changes before you commit something and ask your self: "Is this the change I made? Does it make sense that my change caused this?" In this particular case, you added some function. It would have made sense that this function were added to `NAMESPACE`. Instead almost everything got deleted from  `NAMESPACE`. That cannot be correct, so don't commit it.

Here the error was subtle: You used `library(swephR)` in a file in the `R` sub-directory. Simply put, these files are executed when the package is loaded. But now R is supposed to load the package while loading it. That cannot work. Unfortunately the `devtools` used b RStudio are not good at catching this error. I only saw it when I did an `R CMD INSTALL`. As a general rule: never use `library` or `require` in a file placed into the `R` sub-directory.

vreijs commented:
I want to stress the importance to stay with teh SE method;-) But I will
have a think about this all. The errors are get are simple: sefstars.txt
file not found.
<

test-misc.R:212: failure: Heliacal rise of sirus
result$serr not equal to "".
1/1 mismatches
x[1]: "SwissEph file 'sefstars.txt' not found
x[1]:  in PATH 'C:/Users/VRHome/Documents/R/s
x[1]: wephR/ephemeris\\'"
y[1]: ""
---------------------------------------------

>

Don't know if the insert works (as I use my normal e-mail program (gmail))
for answering.

The multiple directories might make it even more  difficult when manual
intervention is needed;-) Could we do the same tests in R as SE does and
base on that the testing?

vreijs commented:
Thanks for these advises. Will be more careful.

rstub commented:
> By the way the message i get when I run swe_heliacal_ut standalone (and not in ARCHAECOSMO) I get: 'R Session Aborted, R encontered a fatal error. The session was terminated. 'Start New Session?'" 

I have also seen lots of strange segmentation faults since working on this method. It seems that SE does something with the memory of `dobs` that R does not like. As a workaround I have now separated the memory location used by SE and R.

rstub committed:
Add devtools workaround

vreijs reviewed (commented):
I think it might be the dobs that does the error! In my ARCHAEOCOSMO I fill in all arguments (6), but not when directly calling the swe_heliacal_swe: I only used 2 (as I read from SE manual that only three were needed), but then the array was perhaps too small for the C-code call and I can imagine that causes the problem.

rstub committed:
Exclude only at the top-level

rstub commented:
```
x[1]: "SwissEph file 'sefstars.txt' not found 
x[1]: in PATH 'C:/Users/VRHome/Documents/R/s 
x[1]: wephR/ephemeris\\'"
```

Ah, that is probably related to https://github.com/r-lib/devtools/issues/1617. I have tried to fix this in my last commit, but I am not sure if that works on Windows. That really is a problem in `devtools`. If you try 'check' (Ctrl-Shift-E) I am quite sure that the tests run correctly also on your machine.

rstub commented:
> I think it might be the dobs that does the error! In my ARCHAEOCOSMO I fill in all arguments (6), but not when directly calling the swe_heliacal_swe: I only used 2 (as I read from SE manual that only three were needed), but then the array was perhaps too small for the C-code call and I can imagine that causes the problem.

That sounds like a plausible explanation. So either we add tests to ensure that the incomming vectors are long enough, or we do this copying to fixed size arrays for all input vectors. I am not sure yet what I prefer ...

rstub commented:
> The multiple directories might make it even more difficult when manual intervention is needed;-)

I don't think so. This would make it possible to place an extended `sefstars.txt` in some location and just prepend that path to the overall search path.

> Could we do the same tests in R as SE does and base on that the testing?

Does SE have a fixed test set including expected results? I would *love* to have something like that.

vreijs commented:
I think we should fill in the full vector for now: the default can be c(age, SN, 1,0,5,0.8). This should give the same results as c(age,SN) if SE_HELFLAG_OPTICAL_PARAMS  is not set. 
For me it is on the list to query the SE authors (I keep these things noted in the SE manual: swephprg.doc).

vreijs commented:
This thought prompt me to something else. So if someone does not fill in the arrays properly we might get such a 'hanging' error.. Mhhh, but perhaps not much to do (I don't like to check vector-length before calling C). By the way it would be nice to be able to use swe_set_path(NULL) (as per the manual). Perhaps we could add a check if file=NULL, then file=''

vreijs commented:
It has a test, but at this moment I can't find. I have seen it once (and
might even have even it earlier in an earlier 'issues' e-mail). I will see
if I can dig it up!

vreijs commented:
Refound the method of finding ephemeris files (page 67 of swephprg.doc):
<
How the sample programs search for the ephemeris files:

1.	check environment variable SE_EPHE_PATH; if it exists it is used, and if it has invalid content, the program fails.
2.	Try to find the ephemeris files in the current working directory
3.	Try to find the ephemeris files in the directory where the executable resides
4.	Try to find a directory named \SWEPH\EPHE in one of the following three drives: 
•	where the executable resides
•	current drive
•	drive C:

As soon as it succeeds in finding the first ephemeris file it looks for, it expects all required ephemeris files to reside there. This is a feature of the sample programs only, as you can see in our C code. >

I see this is the method of the sample program (not of SE, sorry I was wrong there), but perhaps we can use that also for R?

vreijs commented:
Aha, I now see that you want to refer to a sefstars.txt in the directory:

C:/Users/VRHome/Documents/R/swephR/ephemeris

I have a file 'ephemeris' with content inst/ephemeris and in
inst/ephemeris directory is the sefstars.txt file.

Can you not look directly in
C:/Users/VRHome/Documents/R/swephR/inst/ephemeris' for this test?

All the best,


Victor

vreijs committed:
tests for eclipses and updated test for heliacl event (full vector).

vreijs committed:
restricted comparison to valid array elements

vreijs committed:
commented the swe_set_ephe_path statement (is important for VR's testing

vreijs commented:
I tried to check within swe_set_ephe_path if (file==NULL) but that seems not to be allowed Nul is int and string is an string of char...

rstub committed:
Revert "Exclude only at the top-level"

This reverts commit ccab017dbdc30afb5f2ad3b844240e7d0d917053.

rstub committed:
Revert "Add devtools workaround"

This reverts commit e2162b971acdec98f8bb464f0f37b9f97e430817.

vreijs committed:
added swe_heliacal_pheno_ut and swe_vis_limit_mag

vreijs committed:
updated a few array sizes (matched to SE VBA file)

rstub committed:
Access packaged ephemeris in devtools compatible way

vreijs committed:
removed swe_rise_trans (is vertaken by swe_rise_trans_true_hor

vreijs committed:
this test routine not needed anymore

rstub committed:
Remove WARNING by using consistent argument naming

rstub committed:
Allow NULL as argument to swe_set_ephe_path

rstub committed:
Differentiate test results from swe_pheno_ut

For unknown reasons the results from swe_pheno_ut depend on whether
or not 'swephRdata' is installed, even wenn the build in ephemeris
is used.

rstub commented:
With my latest commits it is now possible to use `swe_set_ephe_path(NULL)`. In addition, it should now be possible for you to test without calling `swe_set_ephe_path("")` first.

One strange thing: the results from `swe_pheno_ut()` depend on whether or not 'swephRdata' is installed, even wenn the build in ephemeris is used. Any idea why that might be the case?

vreijs committed:
included swe_topo_arcus_visionis

vreijs committed:
Merge branch 'feature/vectorization' of https://github.com/rstub/swephR into feature/vectorization

rstub committed:
Use (more) consistent argument naming

rstub committed:
Check size of NumericVector arguments

rstub committed:
Document missing arguments from swe_topo_arcus_visionis()

vreijs commented:
On 8 September 2018 at 21:22, Ralf Stubner <notifications@github.com> wrote:

> With my latest commits it is now possible to use swe_set_ephe_path(NULL).
>

When I call with NULL argument, I get following error:

> swe_set_ephe_path(NULL)Error in swe_set_ephe_path(NULL) :
  Expecting a single string value: [type=NULL; extent=0].



In addition, it should now be possible for you to test without calling
> swe_set_ephe_path("") first.
>
Indeed no need for it anymore.

> One strange thing: the results from swe_pheno_ut() depend on whether or
> not 'swephRdata' is installed, even wenn the build in ephemeris is used.
> Any idea why that might be the case?
>

Should not depend, certainly when explicitly ephemeris=4 is being used.
Perhaps also something to mention to swissephe e-group.
I have not yet installed swephRdata, but will do that soon. Would another
name for this data not be better, like DE431SEdata? This links is very
close to SE instead of swephR...

vreijs commented:
On 8 September 2018 at 21:22, Ralf Stubner <notifications@github.com> wrote:

> One strange thing: the results from swe_pheno_ut() depend on whether or
> not 'swephRdata' is installed, even wenn the build in ephemeris is used.
> Any idea why that might be the case?
>

I get teh same results as you when using

swe_pheno_ut(1234567,1,4)

with sweph3data. So there is something wrong. Indeed to be discussed
at swissephe e-group, IMHO.


—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#issuecomment-419670438>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FlpENd3GPAqiIutkc3uI_hbaD8Btks5uZCcagaJpZM4WZw00>
> .
>

rstub commented:
>When I call with NULL argument, I get following error:
> swe_set_ephe_path(NULL)
> Error in swe_set_ephe_path(NULL) :
> Expecting a single string value: [type=NULL; extent=0].

This is odd. Was that in a new R session? Maybe you did not install the most recent version with "Build & Restart"?

> I get teh same results as you when using
>
> swe_pheno_ut(1234567,1,4)
> 
>with sweph3data. So there is something wrong. Indeed to be discussed at swissephe e-group, IMHO.

ACK. Let's try to reproduce this in plain C first. If we can't do that, then we know there is something wrong with our code. If we can do that, then they will have an easier time to fix it. Do you know the command-line arguments needed to do this sort of calculation with their 'swetest' program?

> Would another name for this data not be better, like DE431SEdata? This links is very close to SE instead of swephR...

Naming things is famously difficult ... I used the swephRdata since the package is primarily meant as data companion to swephR. However, one could indeed use these files with other tools using the SE software. One only has to set the environment variable SE_EPHE_PATH appropriately. How about DE431SE? The trailing 'data' appears superfluous to me.

rstub committed:
Initialize arrays used as output variables

rstub committed:
Use the general pattern to remove the full output array

(fixes #19)

rstub commented:
BTW, once the issue with `swe_set_ephe_path(NULL)` is fixed, I would like to merge this PR. It is getting way to large for my liking. Are there any other loos ends?

vreijs commented:
On 9 September 2018 at 09:24, Ralf Stubner <notifications@github.com> wrote:

> When I call with NULL argument, I get following error:
> swe_set_ephe_path(NULL)
> Error in swe_set_ephe_path(NULL) :
> Expecting a single string value: [type=NULL; extent=0].
>
> This is odd. Was that in a new R session? Maybe you did not install the
> most recent version with "Build & Restart"?
>

I did a 'Clean and Build' did not work, then 'Install and Restart' did not
work
then a roxygenise() and then it worked!
I still don't udnerstand sometimes the workflow needed (although I am
getting better;-).

> I get teh same results as you when using
>
> swe_pheno_ut(1234567,1,4)
>
> with sweph3data. So there is something wrong. Indeed to be discussed at
> swissephe e-group, IMHO.
>
> ACK. Let's try to reproduce this in plain C first. If we can't do that,
> then we know there is something wrong with our code. If we can do that,
> then they will have an easier time to fix it. Do you know the command-line
> arguments needed to do this sort of calculation with their 'swetest'
> program?
>
Will check this later.

> Would another name for this data not be better, like DE431SEdata? This
> links is very close to SE instead of swephR...
>
> Naming things is famously difficult ... I used the swephRdata since the
> package is primarily meant as data companion to swephR. However, one could
> indeed use these files with other tools using the SE software. One only has
> to set the environment variable SE_EPHE_PATH appropriately. How about
> DE431SE? The trailing 'data' appears superfluous to me.
>
I know, naming is sometimes a pain in the neck. Certainly in an operational
environment, if one choice for the wrong naming, one is stuck with it for
years... SE calls the flag SEFLG_SWIEPH. We also could call it SWIEPH...




> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#issuecomment-419699223>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FhTJKABPa0PXVaJZ9wEv7rpmVPS_ks5uZNA0gaJpZM4WZw00>
> .
>

rstub commented:
@vreijs I have now merged this PR. For your next changes please use the branch https://github.com/rstub/swephR/tree/vreijs/functions3.

vreijs commented:
In vreijs/function3 I don't see your NULL condition for set_ephe_path... SI
thsi thus the correct branch?

On 9 September 2018 at 15:30, Ralf Stubner <notifications@github.com> wrote:

> @vreijs <https://github.com/vreijs> I have now merged this PR. For your
> next changes please use the branch https://github.com/rstub/
> swephR/tree/vreijs/functions3.
>

rstub commented:
Did you make a `git pull` in that branch?

rstub commented:
BTW, I have opened a new issue https://github.com/rstub/swephRdata/issues/1 for discussing a possible rename of `swephRdata`.

vreijs commented:
Hello Ralf,

When I do a pull I get:

   cd21f21..61c25cf  vreijs/functions3     -> origin/vreijs/functions3
   ef9f0cb..4865637  feature/vectorization -> origin/feature/vectorization
   3c9b0d8..61c25cf  master                -> origin/master
Updating cd21f21..61c25cf
error: Your local changes to the following files would be overwritten by merge:
	NAMESPACE
	man/expert-interface.Rd
Please commit your changes or stash them before you merge.
Aborting

In some way I just want to just load your version (and forgetting all
my stuff). In the past I deleted the things that were on my machine
and then pull again from your Github, but is there another/better way?
What does 'stash' mean (I assume 'foget' what I have, and move it to
some place for future) and how do I do that? Thanks for your help.

All the best,


Victor


On 9 September 2018 at 16:16, Ralf Stubner <notifications@github.com> wrote:

> Did you make a git pull in that branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#issuecomment-419722868>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6Fq83VEWIQtKuA1IL0QAv2QKHYCU2ks5uZTDlgaJpZM4WZw00>
> .
>

vreijs commented:
At the end I deleted the swephR folder on my PC and did a new project...

On 9 September 2018 at 16:16, Ralf Stubner <notifications@github.com> wrote:

> Did you make a git pull in that branch?
>

rstub commented:
You got the meaning of 'git stash' right. Unfortunately RStudio does not support this within the GUI. You have to use a terminal, which can be done inside RStudio, and call

    git stash

This will store all local changes in a way, that they can be retrieved later on with

    git stash pop

For more info see here: https://git-scm.com/book/en/v2/Git-Tools-Stashing-and-Cleaning

BTW, I think in your branch you are missing a call to 'roxygenize()'. Or you have forgotten to add 'NAMESPACE' to the last commit. At least the CI systems are complaining that they cannot find one of your new functions, which is also not present in the 'NAMESPACE' file. Could you open a PR via the web interface for that branch? This would simplify discussions.

vreijs commented:
On 9 September 2018 at 21:31, Ralf Stubner <notifications@github.com> wrote:

> You got the meaning of 'git stash' right. Unfortunately RStudio does not
> support this within the GUI. You have to use a terminal, which can be done
> inside RStudio, and call
>
> git stash
>
> Will do that next time (and see how it works).

BTW, I think in your branch you are missing a call to 'roxygenize()'.
>
I did many time a roxyginise, as I indeed did't see the swe_calc_ut... But
I am not able to get it there.
I think I had it once chenged:
// Compute information of planet (ET)
// [[Rcpp::export]]
in to
// Compute information of planet (ET)
// [[Rcpp::export(swe_calc)]]
<and similar for swe_calc_ut>
And then I saw swe_calc_ut but not swe_calc...

rstub commented:
Right,  swe_calc  is special in that it does not use `//' @export`. Instead, in https://github.com/rstub/swephR/blob/master/R/expert-interface.R there is a hand written wrapper function that is exported. You will need something like this here as well.

vreijs commented:
Will work on this this evening (now at work).

On 9 September 2018 at 22:04, Ralf Stubner <notifications@github.com> wrote:

> You will need something like this here as well.
>

vreijs commented:
Ok, I now added the expert-interface.R lines, it works now. Thanks. I will
look if I need to add more functions (to make sure that the skyscapeR
package has also the right functions, beside ARCHEAOCOSMO). I am now
analysing the skyscapeR code. I hope to have this done within a few days.

On 9 September 2018 at 22:04, Ralf Stubner <notifications@github.com> wrote:

> Right, swe_calc is special in that it does not use //' @export. Instead,
> in https://github.com/rstub/swephR/blob/master/R/expert-interface.R there
> is a hand written wrapper function that is exported. You will need
> something like this here as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/17#issuecomment-419744943>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FnE6ESsLosQ0ezCppunlHyppKN9fks5uZYJngaJpZM4WZw00>
> .
>

rstub commented:
Great. I see on https://github.com/rstub/swephR/branches that CI is fixed. Can you please open a pull request for that branch? You can do that directly from the above web page. That way we can move the discussion where it belongs.
